File: 0155-min-stack/0155-min-stack.java
class MinStack {
    private class Node{
        int x;
        int min;
        Node next;
        private Node(int x,int min,Node next){
            this.x=x;
            this.min=min;
            this.next=next;
        }
        
    }
    private Node head;
    
    public void push(int val) {
        if(head==null){
            head=new Node(val,val,null);
        }
        else{
            head=new Node(val,Math.min(val,head.min),head);
        }
    }
    
    public void pop() {
        head=head.next;
    }
    
    public int top() {
        return head.x;
    }
    
    public int getMin() {
        return head.min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 0155-min-stack/0155-min-stack.py
import sys
class MinStack:

    def __init__(self):
        self.s=[]
        self.min=sys.maxsize
        

    def push(self, val: int) -> None:
        self.min=min(val,self.min)
        self.s.append([val,self.min])
        

    def pop(self) -> None:
        self.s.pop()
        if self.s:
            self.min=self.s[-1][1]
        else:
            self.min=sys.maxsize

    def top(self) -> int:
        return self.s[-1][0]
        

    def getMin(self) -> int:
        return self.s[-1][1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()